### DISCUSSION TITLE
composite2 with saturate blend mode is not working

### DISCUSSION THREAD

**Discussion by jfcalvo** [libvips/libvips#2250] [Q&A] [answered]
Hi, not sure if this is a bug but something is not working as expected when using `composite2` with `saturate` blend mode.

I'm using the two following images:

![base](https://user-images.githubusercontent.com/446497/117567150-0689d000-b0bb-11eb-8a3f-3c4c497a1c91.jpg)
![overlay](https://user-images.githubusercontent.com/446497/117567154-0a1d5700-b0bb-11eb-9094-387bcf4275f6.jpg)

Using the `darken` blend mode I get expected results:
```sh
$ vips composite2 base.jpg overlay.jpg output_darken.jpg darken
```

![output_darken](https://user-images.githubusercontent.com/446497/117567169-1f928100-b0bb-11eb-9b4d-69a76ea394a1.jpg)

But when using `saturate` blend mode what I get is the original image without any blending applied over it:
```sh
vips composite2 base.jpg overlay.jpg output_saturate.jpg saturate
```

![output_saturate](https://user-images.githubusercontent.com/446497/117567196-49e43e80-b0bb-11eb-8ad2-4be156db55a9.jpg)

Using `vips-8.10.2-Mon Oct 12 16:43:59 UTC 2020` on MacOS `10.14.6`.

Thank you.

**Comment by jcupitt** [comment, marked as answer]
Hi @jfcalvo,

libvips uses the Cairo blend modes, so `saturate` is defined like this:

https://www.cairographics.org/operators/#saturate

You can see for pixel values A and B the result is `x * A + B`, where `x` is `min(alphaA, 1−alphaB)`. For two non transparent pixels (alpha is 1.0), this will always be B. So Cairo `saturate` only affects the blending of semi-transparent pixels.

`darken` is one of the PDF blending modes and they have a pretty different definition, if you read down the page.

**Reply by jfcalvo** [reply]
Understood. Thank you very much for the explanation and the link to the different blend modes.
